### PR TITLE
fix missing UID when update event

### LINKED ISSUE #14126: Events missing UIDs
Certain events are missing UIDs. After a bit of investigation with @lavalamp and @saad-ali we realized that UIDs are not validated if existing events are updated. The api-server can populate UIDs for events if they are missing based on the Name. 
Fixing the client side event generator may not work everywhere since it requires re-compiling all the clients.

### PR DESCRIPTION
**Pull request by jiangyaoguo**
Fixed #14126

### CODE REVIEW

**Comment by jiangyaoguo** [comment]
@vishh @lavalamp @saad-ali

File: pkg/registry/event/strategy.go [hunk 1]
```diff
@@ -59,6 +60,9 @@ func (eventStrategy) AllowCreateOnUpdate() bool {
 
 func (eventStrategy) ValidateUpdate(ctx api.Context, obj, old runtime.Object) fielderrors.ValidationErrorList {
 	event := obj.(*api.Event)
+	if event.ObjectMeta.UID == "" {
+		event.ObjectMeta.UID = util.NewUUID()
```
Comment (thockin) [inline pkg/registry/event/strategy.go]: wou;dn't we want to copy the UID from `old` rather than making a new one?

Comment (jiangyaoguo) [inline pkg/registry/event/strategy.go]: @thockin  You're right. Sorry, I misunderstand what we want. Will fix it tomorrow.

Comment (vishh) [inline pkg/registry/event/strategy.go]: +1. This is mainly for events that get updated and so we need to preserve the UID.

Comment (lavalamp) [inline pkg/registry/event/strategy.go]: hm we really need to test for this generally. No update changing the UID is valid, and if we're going to be nice like this for one resource, we should be nice for all resources.

**Comment by bgrant0607** [comment]
Ping @vishh @saad-ali. Do you want this in 1.1?

**Comment by lavalamp** [comment]
*I* want it in 1.1. This is embarrassing. But is this the right fix? should we be nice like this? how do we know the user intended to modify this thingy as opposed to the one that had this name up until a millesecond ago but was then deleted and this new one created? And does this properly prevent the UID from changing?

**Comment by saad-ali** [comment]
My 2 cents:

I'd like to get a fix in. But I feel it is non-critical: events are uniquely identified by their name by all consumers (@vishh made modifications to the ones that weren't, heapster/GCL, to make sure they are).

The proposed change, modifying the update request on the server side, feels like a bad hack to me. The API server shouldn't get into the business of trying to guess what a client meant. Instead, I'd rather push this logic back to the client library, since it is responsible for [constructing the update request](https://github.com/kubernetes/kubernetes/blob/master/pkg/client/record/event.go#L118). That would be a quick fix, but @vishh mentioned in the original issue that he preferred a solution that didn't require recompiling all clients.

I'll leave it to @vishh to decide.

**Comment by lavalamp** [comment]
I'm assinging this to @caesarxuchao, and also asking him to audit for other UID checking problems. I'm still not sure if we should "help" the user by adding it if it is blank-- but we need to fix all clients before turning on validation if we don't.

**Comment by caesarxuchao** [comment]
I agree with @saad-ali that we shouldn't modify the update request on the server side, especially not in BeforeUpdate, because that would affect resources other than `event`. As Saad pointed out, I think a proper fix would be fixing the client library that generates the update request.

IIUC, as Saad pointed out events have unique names, so "events missing UID" hasn't caused any problem yet, has it? Can we live with this problem until clients with this problem are retired? If so, fixing the client library will be enough for now.

@vishh @saad-ali @lavalamp  could you chime in? Thanks.

**Comment by caesarxuchao** [comment]
Let's move the discussion to the original issue #14126, as we probably needs another way to fix the problem. Sorry @jiangyaoguo. Thank you for your efforts in making this PR.

**Comment by lavalamp** [comment]
@saad-ali @vishh If @caesarxuchao is going to fix this the right way, do we still object to accepting this in the short term? I think it should be OK?

> especially not in BeforeUpdate, because that would affect resources other than event

@caesarxuchao but that's a feature, not a bug! We *want* to fix this for all resources--if they have the same problem, which I expect they might.

**Comment by saad-ali** [comment]
> @saad-ali @vishh If @caesarxuchao is going to fix this the right way, do we still object to accepting this in the short term? I think it should be OK?

If we think it will improve something, then yes. But, I don't see what it buys us. I'm ok with skipping this and doing it the "right way" (client update + server side validation) for 1.2.

That said, I'll defer to @vishh.

**Comment by jiangyaoguo** [comment]
@caesarxuchao Sorry for delayed response because of vacation.  Do we should close this now ? I saw  you had made another pr to fix this.

**Comment by caesarxuchao** [comment]
@jiangyaoguo, no problem. Yes, I'm going to close this one. Thank you.

**Pull request closed without merging**